File: src/assembler/agents-selection.ts
/**
 * Agent selection logic: conditional agents and checklist injection rules.
 *
 * Pure functions with no I/O — consumed by {@link AgentsAssembler}.
 *
 * @module
 */
import type { ProjectConfig, InfraConfig } from "../models.js";
import { hasInterface, hasAnyInterface } from "./conditions.js";

const MD_EXTENSION = ".md";

/** A (targetAgent, checklistFile, condition) tuple for checklist injection. */
export interface ChecklistRule {
  readonly agent: string;
  readonly checklist: string;
  readonly active: boolean;
}

/** Select conditional agent filenames based on project configuration. */
export function selectConditionalAgents(
  config: ProjectConfig,
): string[] {
  return [
    ...selectDataAgents(config),
    ...selectInfraAgents(config),
    ...selectInterfaceAgents(config),
    ...selectEventAgents(config),
  ];
}

/** Build the complete list of checklist injection rules. */
export function buildChecklistRules(
  config: ProjectConfig,
): ChecklistRule[] {
  return [
    ...securityChecklistRules(config.security.frameworks),
    ...apiChecklistRules(config),
    ...devopsChecklistRules(config.infrastructure),
  ];
}

/** Derive marker string from checklist filename. */
export function checklistMarker(checklistFile: string): string {
  const name = checklistFile
    .replace(MD_EXTENSION, "")
    .toUpperCase()
    .replaceAll("-", "_");
  return `<!-- ${name} -->`;
}

function selectDataAgents(config: ProjectConfig): string[] {
  if (config.data.database.name !== "none") {
    return ["database-engineer.md"];
  }
  return [];
}

function selectInfraAgents(config: ProjectConfig): string[] {
  const agents: string[] = [];
  const infra = config.infrastructure;
  if (infra.observability.tool !== "none") {
    agents.push("observability-engineer.md");
  }
  const hasDevops =
    infra.container !== "none" ||
    infra.orchestrator !== "none" ||
    infra.iac !== "none" ||
    infra.serviceMesh !== "none";
  if (hasDevops) {
    agents.push("devops-engineer.md");
  }
  return agents;
}

function selectInterfaceAgents(config: ProjectConfig): string[] {
  if (hasAnyInterface(config, "rest", "grpc", "graphql")) {
    return ["api-engineer.md"];
  }
  return [];
}

function selectEventAgents(config: ProjectConfig): string[] {
  const hasEvents =
    config.architecture.eventDriven ||
    hasAnyInterface(config, "event-consumer", "event-producer");
  if (hasEvents) {
    return ["event-engineer.md"];
  }
  return [];
}

function securityChecklistRules(
  frameworks: readonly string[],
): ChecklistRule[] {
  const hasPrivacy =
    frameworks.includes("lgpd") || frameworks.includes("gdpr");
  return [
    { agent: "security-engineer.md", checklist: "pci-dss-security.md", active: frameworks.includes("pci-dss") },
    { agent: "security-engineer.md", checklist: "privacy-security.md", active: hasPrivacy },
    { agent: "security-engineer.md", checklist: "hipaa-security.md", active: frameworks.includes("hipaa") },
    { agent: "security-engineer.md", checklist: "sox-security.md", active: frameworks.includes("sox") },
  ];
}

function apiChecklistRules(config: ProjectConfig): ChecklistRule[] {
  return [
    { agent: "api-engineer.md", checklist: "grpc-api.md", active: hasInterface(config, "grpc") },
    { agent: "api-engineer.md", checklist: "graphql-api.md", active: hasInterface(config, "graphql") },
    { agent: "api-engineer.md", checklist: "websocket-api.md", active: hasInterface(config, "websocket") },
  ];
}

function devopsChecklistRules(infra: InfraConfig): ChecklistRule[] {
  return [
    { agent: "devops-engineer.md", checklist: "helm-devops.md", active: infra.templating === "helm" },
    { agent: "devops-engineer.md", checklist: "iac-devops.md", active: infra.iac !== "none" },
    { agent: "devops-engineer.md", checklist: "mesh-devops.md", active: infra.serviceMesh !== "none" },
    { agent: "devops-engineer.md", checklist: "registry-devops.md", active: infra.registry !== "none" },
  ];
}

File: java/src/main/java/dev/iadev/assembler/AgentsSelection.java
package dev.iadev.assembler;

import dev.iadev.model.InfraConfig;
import dev.iadev.model.ProjectConfig;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Pure agent selection functions based on project config
 * feature gates.
 *
 * <p>These functions evaluate config conditions and return
 * agent filenames. No file I/O — consumed by
 * {@link AgentsAssembler} for assembly decisions.</p>
 *
 * <p>Selection categories:
 * <ul>
 *   <li>Data agents — database-engineer when database
 *       is configured</li>
 *   <li>Infrastructure agents — observability-engineer,
 *       devops-engineer based on infra config</li>
 *   <li>Interface agents — api-engineer when REST, gRPC,
 *       or GraphQL interfaces present</li>
 *   <li>Event agents — event-engineer when event-driven
 *       or event interfaces present</li>
 * </ul>
 *
 * <p>Example usage:
 * <pre>{@code
 * List<String> conditionals =
 *     AgentsSelection.selectConditionalAgents(config);
 * String devAgent =
 *     AgentsSelection.selectDeveloperAgent(config);
 * }</pre>
 * </p>
 *
 * @see AgentsAssembler
 * @see ConditionEvaluator
 */
public final class AgentsSelection {

    private static final String MD_EXTENSION = ".md";

    private AgentsSelection() {
        // utility class
    }

    /**
     * Evaluates all feature gates and returns conditional
     * agent filenames.
     *
     * @param config the project configuration
     * @return aggregated list of conditional agent filenames
     */
    public static List<String> selectConditionalAgents(
            ProjectConfig config) {
        List<String> agents = new ArrayList<>();
        agents.addAll(selectDataAgents(config));
        agents.addAll(selectInfraAgents(config));
        agents.addAll(selectInterfaceAgents(config));
        agents.addAll(selectEventAgents(config));
        return agents;
    }

    /**
     * Returns the developer agent filename for the project
     * language.
     *
     * @param config the project configuration
     * @return developer agent filename (e.g. "java-developer.md")
     */
    public static String selectDeveloperAgent(
            ProjectConfig config) {
        String langName = config.language().name();
        return langName + "-developer" + MD_EXTENSION;
    }

    /**
     * Builds the complete list of checklist injection rules.
     *
     * @param config the project configuration
     * @return list of checklist rules with active flags
     */
    public static List<ChecklistRule> buildChecklistRules(
            ProjectConfig config) {
        List<ChecklistRule> rules = new ArrayList<>();
        rules.addAll(securityChecklistRules(
                config.security().frameworks()));
        rules.addAll(apiChecklistRules(config));
        rules.addAll(devopsChecklistRules(
                config.infrastructure()));
        return rules;
    }

    /**
     * Derives the marker string from a checklist filename.
     *
     * <p>Example: "pci-dss-security.md" becomes
     * "&lt;!-- PCI_DSS_SECURITY --&gt;"</p>
     *
     * @param checklistFile the checklist filename
     * @return the HTML comment marker string
     */
    public static String checklistMarker(
            String checklistFile) {
        String name = checklistFile
                .replace(MD_EXTENSION, "")
                .toUpperCase()
                .replace("-", "_");
        return "<!-- " + name + " -->";
    }

    private static List<String> selectDataAgents(
            ProjectConfig config) {
        if (!"none".equals(
                config.data().database().name())) {
            return List.of("database-engineer.md");
        }
        return List.of();
    }

    private static List<String> selectInfraAgents(
            ProjectConfig config) {
        List<String> agents = new ArrayList<>();
        InfraConfig infra = config.infrastructure();

        if (!"none".equals(
                infra.observability().tool())) {
            agents.add("observability-engineer.md");
        }

        boolean hasDevops =
                !"none".equals(infra.container())
                        || !"none".equals(
                                infra.orchestrator())
                        || !"none".equals(infra.iac())
                        || !"none".equals(
                                infra.serviceMesh());
        if (hasDevops) {
            agents.add("devops-engineer.md");
        }
        return agents;
    }

    private static List<String> selectInterfaceAgents(
            ProjectConfig config) {
        if (hasAnyInterface(config,
                "rest", "grpc", "graphql")) {
            return List.of("api-engineer.md");
        }
        return List.of();
    }

    private static List<String> selectEventAgents(
            ProjectConfig config) {
        boolean hasEvents =
                config.architecture().eventDriven()
                        || hasAnyInterface(config,
                                "event-consumer",
                                "event-producer");
        if (hasEvents) {
            return List.of("event-engineer.md");
        }
        return List.of();
    }

    private static List<ChecklistRule> securityChecklistRules(
            List<String> frameworks) {
        boolean hasPrivacy = frameworks.contains("lgpd")
                || frameworks.contains("gdpr");
        return List.of(
                new ChecklistRule(
                        "security-engineer.md",
                        "pci-dss-security.md",
                        frameworks.contains("pci-dss")),
                new ChecklistRule(
                        "security-engineer.md",
                        "privacy-security.md",
                        hasPrivacy),
                new ChecklistRule(
                        "security-engineer.md",
                        "hipaa-security.md",
                        frameworks.contains("hipaa")),
                new ChecklistRule(
                        "security-engineer.md",
                        "sox-security.md",
                        frameworks.contains("sox")));
    }

    private static List<ChecklistRule> apiChecklistRules(
            ProjectConfig config) {
        return List.of(
                new ChecklistRule(
                        "api-engineer.md",
                        "grpc-api.md",
                        hasInterface(config, "grpc")),
                new ChecklistRule(
                        "api-engineer.md",
                        "graphql-api.md",
                        hasInterface(config, "graphql")),
                new ChecklistRule(
                        "api-engineer.md",
                        "websocket-api.md",
                        hasInterface(config, "websocket")));
    }

    private static List<ChecklistRule> devopsChecklistRules(
            InfraConfig infra) {
        return List.of(
                new ChecklistRule(
                        "devops-engineer.md",
                        "helm-devops.md",
                        "helm".equals(infra.templating())),
                new ChecklistRule(
                        "devops-engineer.md",
                        "iac-devops.md",
                        !"none".equals(infra.iac())),
                new ChecklistRule(
                        "devops-engineer.md",
                        "mesh-devops.md",
                        !"none".equals(infra.serviceMesh())),
                new ChecklistRule(
                        "devops-engineer.md",
                        "registry-devops.md",
                        !"none".equals(infra.registry())));
    }

    private static boolean hasInterface(
            ProjectConfig config, String type) {
        return config.interfaces().stream()
                .anyMatch(i -> type.equals(i.type()));
    }

    private static boolean hasAnyInterface(
            ProjectConfig config, String... types) {
        Set<String> typeSet = Set.of(types);
        return config.interfaces().stream()
                .anyMatch(i -> typeSet.contains(i.type()));
    }

    /**
     * A (targetAgent, checklistFile, condition) tuple for
     * checklist injection.
     *
     * @param agent     the agent file to inject into
     * @param checklist the checklist file to inject
     * @param active    whether this rule is active
     */
    public record ChecklistRule(
            String agent,
            String checklist,
            boolean active) {
    }
}
